### DISCUSSION TITLE
Updates in consideration

### DISCUSSION THREAD

**Discussion by occupant** [ubc-web-services/ubc_d8_config_modules#6] [General]
Since it's been deployed a few times and generally working, these are some suggested alterations / additions for refining this project further.

## Disable
- [x] Workflows (complicates and confuses editing experience, especially with node Revisions on)

## Remove
- [ ] image_widget_crop (too situation-specific and requires too much config)
- [ ] Slideshow Paragraph (js library in use is problematic in practice, needs a rethink)

### Asset Library / Media settings
- [x] [Media Bulk Upload](https://www.drupal.org/project/media_bulk_upload) (adds bulk upload options)
- [x] [Media Entity File Replace](https://www.drupal.org/project/media_entity_file_replace) (adds ability to replace file instead of renaming as _0)
- [x] [Linkit Media Library](https://www.drupal.org/project/linkit_media_library) (adds Media Library button to Link wysiwyg button popup, for inserting links to docs)

### Editing tweaks / enhancements
- [x] [Focal Point](https://www.drupal.org/project/focal_point) (instead of Manual Crops - allows for simpler editing experience and more latitude than image_widget_crop, easier to configure)
- [x] [Formtips](https://www.drupal.org/project/formtips) (tooltips on hover / click)
- [x] [Maxlength](https://www.drupal.org/project/Maxlength) (set field input length with realtime status)
- [x] [Optional End Date](https://www.drupal.org/project/optional_end_date) (allows dates to be handled easier in a single field - extends daterange)
- [x] [Scheduler](https://www.drupal.org/project/scheduler) (provides scheduled node publication)
- [x] [Text Summary Options](https://www.drupal.org/project/text_summary_options) (exposes settings to require / always show the summary field)
- [x] [Editor Advanced Link](https://www.drupal.org/project/editor_advanced_link) (adds some additional options to Link wysiwyg button popup - used to add Target)
- [ ] [Editor Button Link](https://www.drupal.org/project/editor_button_link) (integrate button classes in Link wysiwyg button)
- [x] [Chosen](https://www.drupal.org/project/chosen) (Admin and public facing, allows selects to be searchable, themeable - does make jQuery a requirement though)
- [ ] Choose an updated admin theme - Drupal 9 is using [Claro](https://www.drupal.org/project/claro) and there are interesting themes like [Gin](https://www.drupal.org/project/gin) that declutter the interface. 
- [ ] [Responsive Table Filter](https://www.drupal.org/project/responsive_table_filter) (adds a figure element around tables in wysiwyg and adds horizontal scroll)

### Other
- [x] [Google Tag Manager](https://www.drupal.org/project/google_tag) (add by default, but leave it disabled)
- [x] [Redirect](https://www.drupal.org/project/redirect) (missed this one)

## Reassess
- [ ] Core Media vs image field - for fielded images, we have traditionally used the image field instead of allowing media. Core now recommends against this practice as Media support is in core. Primary considerations are enforcing dimensions, crops, focal points, all of which may be solved. Need to examine.

### Events
- [ ] recurring date function (eg, every wednesday during may at 6pm)
- [x] ongoing flag (instead of a start-end date)
- [ ] Views to accomodate these with upcoming events
- [ ] Add to Calendar module (eg Allard) with config / options

### Custom blocks
- [x] Twitter Stream embed
- [ ] Instagram embed

### Image gallery library and styles
- [ ] Vue, Tinyslider (Sauder)?

### WYSIWYG
- [ ] iFrame (ckeditor_iframe module is essentially abandoned. Do we roll our own?, Use Media type, other?)

### Sitewide alerts
- [x] start with Security.ubc.ca implementation?

### Automated testing
- [ ] add testing framework once we decide on a way forward. [Nightwatch](https://nightwatchjs.org) is in core but requires selenium. If we want to use a lighter, javascript-based framework, [cypress](https://www.cypress.io) could be a good candidate.

## Consider
- [ ] [WebP image format](https://www.drupal.org/project/webp) (assuming p.sh can support)

### UBC Content Items
- Custom module not currently D9 compatible - 19 errors, 3 warnings
- Errors
  - [ ] Call to deprecated method entityManager()
  - [ ] Call to deprecated function format_date()
  - [ ] Call to deprecated method link()
  - [ ] Call to deprecated function drupal_set_message()
  - [ ] Call to deprecated constant REQUEST_TIME
- Warnings
  - [ ] .info version requirement
  - [ ] Class PHPUnit\Framework\TestCase not found

**Comment by occupant** [comment]
Review and assign these, creating a new issue for each as they are accepted as changes to make. Tag the new issues with [Changes September 2020 project](https://github.com/ubc-web-services/ubc_d8_config_modules/projects/1) to populate Kanban board and track.

**Comment by occupant** [comment]
EDIT - added test note above

**Comment by ficklecat** [comment]
> * [ ]  [Editor Button Link](https://www.drupal.org/project/editor_button_link) (integrate button classes in Link wysiwyg button)
> * [x]  [Chosen](https://www.drupal.org/project/chosen) (Admin and public facing, allows selects to be searchable, themeable - does make jQuery a requirement though)

The Editor Button Link module seems a bit sketchy as has low use and no security support. 
Perhaps there is an alternate to Chosen that doesn't use JQuery. It seems a bit dated now

**Reply by gocaps** [reply]
Did a fairly extensive search for a non jQuery alternative, but didn't come up with anything stable.

**Comment by occupant** [comment]
Agreed, on both counts. 

RE: **Chosen** - I haven't been able to find a vanilla js alternative to Chosen in contrib. I've seen a few possible alternatives in the wild, but that means I'd need to adopt it and maintain it. To date I've been limiting the use of Chosen to the admin area, as that will load jQuery anyway (depending on the admin theme) and it's great for the lists, tags, etc.

RE: **Editor Button Link** - the functionality from this is a huge improvement in UX for the wysiwyg. It adds optional selects to the Link dialog that allow you choose a series of styles to attach. I use it to:

- define the link structure (unstyled, Button, Full Width Button)
- define the text size (sm, md, lg, etc)
- define the specific style (light, dark, etc)

This has allowed me to move away from editor 'styles' - the method used in Drupal 7 no longer works correctly in Drupal 8's ckeditor implementation. The style menu doesn't accept multiple styles on an element. All styles are replaced instead of a single style added / removed. For example, if you want to set up values for all three of the link settings above, you will have to create a single line including ALL classes for every combination. It works, but it's horrible to read and use and you're making styling into a second step that takes place somewhere else.

I'd be open to alternative approaches / modules, but the current Drupal core implementation is very poor.

**Reply by ficklecat** [reply]
Yeah strange how link handling still seems really bad. I was going to suggest LinkIt but it seems they removed support for link attributes in the latest version for some reason 🤷🏼‍♂️ 

I can't find a good alternative to Chosen either which is a shame.
